datasource/schema/schema.go

```diff
@@ -122,6 +125,130 @@ func (s Schema) TypeAtTerraformPath(ctx context.Context, p *tftypes.AttributePat
+	var diags diag.Diagnostics
+
+	// Raise error diagnostics when data source configuration uses reserved
+	// field names for root-level attributes.
+	reservedFieldNames := map[string]struct{}{
+		"connection":  {},
+		"count":       {},
+		"depends_on":  {},
+		"lifecycle":   {},
+		"provider":    {},
+		"provisioner": {},
+	}
+
+	attributes := s.GetAttributes()
+
+	for k, v := range attributes {
+		if _, ok := reservedFieldNames[k]; ok {
+			diags.AddAttributeError(
+				path.Root(k),
+				"Schema Using Reserved Field Name",
+				fmt.Sprintf("%q is a reserved field name", k),
+			)
+		}
+
+		d := validateAttributeFieldName(path.Root(k), k, v)
+
+		diags.Append(d...)
+	}
+
+	blocks := s.GetBlocks()
+
+	for k, v := range blocks {
+		if _, ok := reservedFieldNames[k]; ok {
+			diags.AddAttributeError(
+				path.Root(k),
+				"Schema Using Reserved Field Name",
+				fmt.Sprintf("%q is a reserved field name", k),
+			)
+		}
+
+		d := validateBlockFieldName(path.Root(k), k, v)
+
+		diags.Append(d...)
+	}
+
+	return diags
+}
+
+// validFieldNameRegex is used to verify that name used for attributes and blocks
+// comply with the defined regular expression.
+var validFieldNameRegex = regexp.MustCompile("^[a-z0-9_]+$")
+
+// validateAttributeFieldName verifies that the name used for an attribute complies with the regular
+// expression defined in validFieldNameRegex.
+func validateAttributeFieldName(path path.Path, name string, attr fwschema.Attribute) diag.Diagnostics {
+	var diags diag.Diagnostics
+
+	if !validFieldNameRegex.MatchString(name) {
+		diags.AddAttributeError(
+			path,
+			"Invalid Schema Field Name",
+			fmt.Sprintf("Field name %q is invalid, the only allowed characters are a-z, 0-9 and _. This is always a problem with the provider and should be reported to the provider developer.", name),
+		)
+	}
+
+	if na, ok := attr.(fwschema.NestedAttribute); ok {
+		nestedObject := na.GetNestedObject()
+
+		if nestedObject == nil {
+			return diags
+		}
+
+		attributes := nestedObject.GetAttributes()
+
+		for k, v := range attributes {
+			d := validateAttributeFieldName(path.AtName(k), k, v)
+
+			diags.Append(d...)
+		}
+	}
+
+	return diags
+}
+
+// validateBlockFieldName verifies that the name used for a block complies with the regular
+// expression defined in validFieldNameRegex.
+func validateBlockFieldName(path path.Path, name string, b fwschema.Block) diag.Diagnostics {
+	var diags diag.Diagnostics
+
+	if !validFieldNameRegex.MatchString(name) {
+		diags.AddAttributeError(
+			path,
+			"Invalid Schema Field Name",
+			fmt.Sprintf("Field name %q is invalid, the only allowed characters are a-z, 0-9 and _. This is always a problem with the provider and should be reported to the provider developer.", name),
+		)
+	}
+
+	nestedObject := b.GetNestedObject()
+
+	if nestedObject == nil {
+		return diags
+	}
+
+	blocks := nestedObject.GetBlocks()
+
+	for k, v := range blocks {
+		d := validateBlockFieldName(path.AtName(k), k, v)
```

Comment:
Similarly for block pathing and block nesting modes.

Reply:
https://github.com/hashicorp/terraform-plugin-framework/issues/574